### ISSUE TITLE
Redirect to file fails!

### ISSUE THREAD

**Issue by ololoe** [Drekin/win-unicode-console#23]
from __future__ import unicode_literals
    import win_unicode_console
    win_unicode_console.enable()
    print u"ÄäБб"

Works if it's called from console, but as far as it's redirected to a file:

    print u"ÄäБб"
    UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-3: ordinal not in range(128)

**Comment by anthrotype** [comment]
That's a feature, not a bug. When redirected to a file, win_unicode_console does not replace the standard streams, since the stream is no longer a "tty" interactive console.
You can use PYTHONIOENCODING environment variable to force stdin/stderr/stdout use a different encoding of your choice when redirecting to a file.
https://docs.python.org/2/using/cmdline.html

**Comment by ololoe** [comment]
IMHO that's a shit feature. Users want the module to solve all sdt* related unicode errors - but the module only does half of it. And how should I fix it? Previously I had `sys.stdout = codecs.getwriter('utf8')(sys.stdout)`, would it conflict with this module?

**Comment by Drekin** [comment]
It is not possible to easily solve all std* related Unicode errors. Something should stay error, because it is error. If you redirect a file to Python you are basically asking Python to interpret its content as a string. But the file is just a sequence of bytes. To interpret it as a string, it's up to you to provide its encoding. Python uses some default encoding. If you want to change this default encoding, set PYTHONIOENCODING as @anthrotype suggests. File redirection works ok in Python at it has nothing to do with this package, which solves the Unicode problems with interaction between Python and Windows console.

**Comment by ololoe** [comment]
I want to do it from the program, understand? I don't want to tell user which environment variables he should set.

>it has nothing to do with this package

Except:
    sys.stdout = codecs.getwriter('utf8')(sys.stdout)
    import win_unicode_console
    win_unicode_console.enable()

    D:\Program Files\programming\Python27\lib\site-packages\win_unicode_console\__in
it__.py:27: RuntimeWarning: sys.stdin.encoding == 'utf-8', whereas sys.stdout.en
    coding == None, readline hook consumer may assume they are the same
      readline_hook.enable(use_pyreadline=use_pyreadline)

How to fix it?

**Comment by Drekin** [comment]
`sys.stdout = codecs.getwriter('utf8')(sys.stdout)` – don't do this. If you want to create a UTF-8 encoded file, just `open("your_output_file", encoding="utf-8")` and write to it. If you are just printing Unicode and a user of your program wants to redirect the output to a file, it's up to him to set up his environment. Actually you shouldn't even enable `win_unicode_console` in your script, it's up to its user to fix his Python environment.

The thing I may fix is that warning for the case where the redirection occurs. But I have to find out what consumers of readlinehook do when `sys.std*.encoding == None` (which is the case only under Python 2).

**Comment by ololoe** [comment]
>it's up to him to set up his environment.

Not on windows, dude. Someone already mentioned it in another thread. Windows is not lunix. Users don't have to fix broken software - and it **crashes** if they don't fix it.

>The thing I may fix is that warning for the case where the redirection occurs.

Do it, please.

**Comment by Drekin** [comment]
I don't understand. If they install Python and try to print unicode, it doesn't work. And for Python 2 nothing can be done about that, so they have to fix it by themselves, it won't magically fix itself. They have at least install a package like `win-unicode-console` and configure their Python installation. An it's not such a big deal.

Where exactly what crashes?

**Comment by ololoe** [comment]
No, in my opinion it (can be|is) done by a developer. I.e. software should be made to work with unicode on every system even in python 2. There should be a way to do this.

> They have at least install a package like win-unicode-console

It can be installed by setup script.

>An it's not such a big deal.

In my eyes (of Windows' user) - yes, it is. And what's the problem to choose  utf8 by default?

>Where exactly what crashes? 

    UnicodeEncodeError: 'ascii' codec can't encode characters in position
It's a crash in user's eyes.

**Comment by Drekin** [comment]
So your user is incapable of installing a Python package and chaging `sitecustomize` or changing an environment variable, but he knows how to redirect an output of a program to a file in terminal? Why doesn't your program just create a file in a standard way when needed?

Regarding the crash, as I explainded before, this isn't related to Windows console and so is out of scope of this particular package, similarly to providing unicode `sys.argv` in Python 2. So even if I implemented this, it would be in a separate packages. Again, are you sure no such package already exists? Of course, such packages could be grouped to a universal Windows Unicode Python problems solving bundle, but `win-unicode-console` is not this bundle.

**Comment by ololoe** [comment]
>So your user is incapable of installing a Python package and chaging sitecustomize or changing an environment variable, but he knows how to redirect an output of a program to a file in terminal?

Why so surprised? Have you ever seen an experienced windows user? Redirection is the same for all programs, python specific shit - no.

>Why doesn't your program just create a file in a standard way when needed?

Because I want it to work like it works now - print to stdout with redirect support (Redirect **support**. Sounds funny, doesn't it?)

> this isn't related to Windows console and so is out of scope of this particular package

This is related to std* fix. You made only a half.

>So even if I implemented this, it would be in a separate packages.

If you want...

>Again, are you sure no such package already exists? 

Again, never saw it.

>Of course, such packages could be grouped to a universal Windows Unicode Python problems solving bundle, 

Good idea.

>win-unicode-console is not this bundle.

So you only solve a very small part of unicode problems on windows or python 2.

As a python programmer that targets primarily windows I just want to know - is python really usable under windows? Can a windows user feel himself as 1st class citizen (this includes that all features that are available under lunix are available under windows too without bundle of workarounds **of course** incompatible with each other)? Or should I abandon python then - it will never work properly? Currently, I can't say "yes, it's easily possible". Btw, unicode in the console works in C# out-of-the-box since ages.

**Comment by Drekin** [comment]
>>So your user is incapable of installing a Python package and chaging sitecustomize or changing an environment variable, but he knows how to redirect an output of a program to a file in terminal?

>Why so surprised? Have you ever seen an experienced windows user? Redirection is the same for all programs, python specific shit - no.

Shouldn't an experienced Windows user be able to set an environment variable? Shouldn't he be able to follow instructions, run a few commands in terminal a create a text file at a given location?

>> this isn't related to Windows console and so is out of scope of this particular package

> This is related to std* fix. You made only a half.

Depends how you define the whole. I'm not claming that I've solved all Unicode std* related problems. Regarding the redirection, how does it work on Linux? I think the situation is the same, an encoding has to be provided somehow. So this redirection stuff is not Windows specific.

>>win-unicode-console is not this bundle.

> So you only solve a very small part of unicode problems on windows or python 2.

I've solved just one particular problem I was interested in, and some other problems needed to be solved in order to solve the main problem. To persuade the core devs that my approach is viable, I made this package so other people can use it and share their experience. I'm not particularly interested in Python 2, I just got an idea how `win-unicode-console` could be ported to Python 2, so I've done it.

> As a python programmer that targets primarily windows I just want to know - is python really usable under windows? Can a windows user feel himself as 1st class citizen (this includes that all features that are available under lunix are available under windows too without bundle of workarounds **of course** incompatible with each other)? Or should I abandon python then - it will never work properly? Currently, I can't say "yes, it's easily possible". Btw, unicode in the console works in C# out-of-the-box since ages.

If you ask this question you should try to find the answer in Python 3 rather than Python 2. For me it's viable now: 1) Install Python 3. 2) Install `win-unicode-console` 3) configure `sitecustomize` 4) set `PYTHONIOENCODING` to UTF-8.

**Comment by ololoe** [comment]
>Shouldn't an experienced Windows user be able to set an environment variable? 

It's harder and unneeded since it can be solved by the developer.

>  I think the situation is the same, an encoding has to be provided somehow.

Well, why not select utf-8 by default? What's the problem of doing it?

> If you ask this question you should try to find the answer in Python 3 rather than Python 2

Well, there is lots of legacy shit that **never** will be ported (take a look on XBMC/Kodi plugins, for example).

> 3) configure sitecustomize 4) set PYTHONIOENCODING to UTF-8.

We are moving in circle. My opinion (and not only mine) is that the developer should take care of it. What are the drawbacks of doing so?

**Comment by Drekin** [comment]
>>  I think the situation is the same, an encoding has to be provided somehow.

> Well, why not select utf-8 by default? What's the problem of doing it? Why crashing instead?

Don't ask me, ask Python developers. Probably somebody already asked, just find the answer. So how is it on Linux, is the Python default encoding UTF-8 there? Or is it locale specific, in which case it is the same as on Windows?

>> If you ask this question you should try to find the answer in Python 3 rather than Python 2

> Well, there is lots of legacy shit that **never** will be ported (take a look on XBMC/Kodi plugins, for example).

But that shouldn't affect the answer to your question.

>> 3) configure sitecustomize 4) set PYTHONIOENCODING to UTF-8.

> We are moving in circle. My opinion (and not only mine) is that the developer should take care of it. What are the drawbacks of doing so?

The Python developers should take care of it, not every developer of every particular script. I don't know about any drawbacks, there may be something with backward compatibility. In any case it needs time and people experienced in this domain.

> Is it so that encoding is lunix-related shit because console has **encoding** under lunix while windows console accepts unicode? Even then, almost all lunixes now have utf8 as encoding.

I think that Windows console has also encoding, but it can also directly accept Unicode via `WriteConsoleW API`. You can blame Windows for not being UTF-8 based.

**Comment by ololoe** [comment]
>You can blame Windows for not being UTF-8 based.
= "You can blame Windows for not being lunix"? No, it isn't, so what?

**Comment by Drekin** [comment]
It's not about being Linux, UTF-16 has no advantage over UTF-8. But given the situation, why do you expect Python to use UTF-8 by default for interprocess communication on Windows?

**Comment by ololoe** [comment]
I want it not to fail as far as non-ascii symbol is printed. It doesn't really matter which encoding is chosen.

**Comment by mic-e** [comment]
Just my 2 cents on the issue:

As a Linux person, I'm used to all output streams, especially those of console programs, to simply working and using `utf-8`. If people for some reason have an ancient distro that doesn't use an UTF8 locale, they'll get exceptions, but that's their problem.

I'm not an expert on (nor fan of) Windows and don't know in which particular way stuff is broken there, but it seems to be broken thoroughly. Which is why I'm using your package, which (along with `pyreadline` and `colorama`) "magically" makes Windows output streams a sane environment. Thank you very much for your effort, BTW, I really appreciate it.

However, all this magic sanity sadly disappears as soon as the output is redirected to a file: Warnings are printed, and on unicode output the program fails.

This disruption of sanity is an inconsistency in **my** use case of your package, "Linux person who wants their stuff to simply work as usual even on Windows". Personally, without having evaluated any evidence, I believe that this is the most common use case for your package.

**Comment by Drekin** [comment]
I thought that the most common use case is just when a Windows user wants to print and/or input Unicode in console.

Regarding your use case, doesn't setting `PYTHONIOENCODING` to UTF-8 work for you?

**Comment by mic-e** [comment]
It doesn't seem so, but I'm not sure why. On Windows, I ship the program packaged as a single `.exe` file with `pyinstaller`; this is my `console_init` function:

    def console_init():
        import sys, locale, os
        if os.name == 'nt':
            import win_unicode_console
            win_unicode_console.enable()
        import readline
        import colorama
        colorama.init()
        print(sys.stdout.encoding)
        print(os.environ["PYTHONIOENCODING"])
        print(chr(9786))

When setting the PYTHONENCODING variable and running the exe, I'm getting
    ...\win_unicode_console\__init__.py:27: RuntimeWarning: sys.stdin.encoding == 'utf-8', whereas sys.stdout.encoding == 'cp1252', readline hook consumer may assume they are the same
    cp1252
    utf_8
    Traceback (most recent call last):
    (...)
    File "c:\python34\lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]

When setting the environment variable and manually running my package, I'm getting the correct UTF-8 output, and `sys.stdout.encoding` is correctly printed as `utf_8`, but your warning still appears.

I suppose the former issue is a bug (feature?) in `pyinstaller`...

**Comment by Drekin** [comment]
That is weird. Can you extract some kind of minimal non-working example? When I try a simple script `echo.py`

    import win_unicode_console as wuc
    wuc.enable()
    print(input())

with `PYTHONIOENCODING=utf-8`, then the redirection semms to work:

    py echo.py > foo.txt
    py echo.py < foo.txt
    py echo.py | py echo.py

**Comment by shamashii** [comment]
Hello, I have similar problem:
I'm using win32 py2.7 with  win-unicode-console to do some image processing scripts,
say test.py, usage: test.py namewithunicodechars.bmp  (e,g 🚲.bmp)
It worked like magic when enable
win_unicode_console.enable()
win_unicode_console.enable(use_unicode_argv=True)
except: Drag and drop directly to test.py
It's seemed like the orginal python phraser isn't a "console", so failed due to the "feature(anthrotype commented on 20 Nov 2015)"

Currently, I'm using a bat:
`set curdir=%~dp0
cd /d %curdir%

pillo.py %1`
to deal with the case.
Hope you can do some fix on it if possible.

**Comment by Drekin** [comment]
@shamashii This doesn't seem to be connected with file redirection. I have created a separate issue: #31.

**Comment by ololoe** [comment]
PYTHONIOENCODING is intended for lunix systems where it should be set up correctly in order to print text into console. While using correct api functions, windows doesn't need encodings to be set up for correct console output. That's not features, that's **linux problems** python developers try to port to windows. Don't you think you should't support them?

**Comment by Drekin** [comment]
Well, `PYTHONIOENCODING` generally dictates the encoding of `sys.std*` objects, which may be usefull even in Windows. AFAIK when you redirect the content of a text file to Python the encoding of the file should match the encoding of `sys.stdin` and if you encode you plaintext files in UTF-8, using `PYTHONIOENCODING` is a way to tell that to Python in Windows.

**Comment by ololoe** [comment]
You can't output unicode with any `sys.std*` encoding, and if you use unicode console io functions, they use only one encoding - unicode. And file output encoding should be set by software, not user. Again, lunix problems, not features are ported.

**Comment by Drekin** [comment]
I would argue that at this context Unicode is not an encoding. Roughly, encoding is a description how a subset of Unicode codepoints is turned into a sequence of bytes.

**Comment by ololoe** [comment]
Why do windows users need `PYENCODING` environment variable? Why should it be set for every windows? What's the difference between them? If you'll speak about console encoding - as I mentioned above, you can't output unicode with any console encoding. What else should it be used for?

**Comment by Drekin** [comment]
I don't claim every Windows user needs it. But you claim it is good for nothing and I replied with what I though was a counterexample – it is a way to tell Python in which encoding is the content of an input or output file that is being redirected to stdin and from stdout, respectively.

I agree that it doesn't help with interactive console at all and I think that nobody claims that.

**Comment by ololoe** [comment]
> it is a way to tell Python in which encoding is the content of an input or output file that is being redirected to stdin and from stdout, respectively.

Have you ever seen this in windows software?

**Comment by Drekin** [comment]
How does it matter? I don't claim it is a good idea or that it is massively used, I just claim it's an example of how can that environment variable be used – it may make sense locally when you customize your own environment and use file redirection manually in cmdline.

**Comment by ololoe** [comment]
I just tried `set PYTHONIOENCODING=utf-8`. Python doesn't even output valid utf-8.

**Comment by Drekin** [comment]
When I try `py -c "print('α')" > out.txt`, it saves as utf-8 when `PYTHONIOENCODING=utf-8` is set. If you are talking about output to console, then of course it doesn't work, that's why `win_unicode_console` exists in the first place.

**Comment by ololoe** [comment]
This doesn't work with 2.7.

**Comment by Drekin** [comment]
`py -2 -c "print(unichr(0x3b1))" > out.txt` works, you cannot use “α” directly because of another issue that cripples non-ASCII command line arguments (see #20).

**Comment by ololoe** [comment]
LOL!!!!!! Fuck python 2! Never thought this bug would come out here.

Btw, `py` starts python 2 if available, you have to explicitly use `py -3`  if you need python 3.

**Comment by Drekin** [comment]
Oh, sorry, I have `PY_PYTHON=3` since the `py` launcher came out :-). Also this default changes with Python 3.6 (https://docs.python.org/3.6/whatsnew/3.6.html).

**Comment by ololoe** [comment]
Where was it changed? I can't find it.

**Comment by Drekin** [comment]
https://docs.python.org/3.6/whatsnew/3.6.html#summary-release-highlights, the “Windows improvements” section.

**Comment by ololoe** [comment]
In any case it's system dependent.

**Comment by Drekin** [comment]
Do you mean the configuration of default Python version or the fact that the `py` launcher is Windows specific?

**Comment by ololoe** [comment]
Which python branch `py` starts.

**Comment by Drekin** [comment]
Not necessarily, `py` means exactly “use the default Python”, you may use it if you want to convey this particular meaning or if you want to stress that it doesn't matter which version is used (which was not the case with that my comment; I have edited that) or when the Python version meant is clear from context. It is similar to writing just `log` for logarithm with implied base.

**Comment by ololoe** [comment]
Yeah, and "default python" on windows is system dependent  (=different for every pc).

https://github.com/Drekin/win-unicode-console/issues/23#issuecomment-241848063 can it be fixed, or is there no hope?

**Comment by Drekin** [comment]
I think there is little hope to have it fix in core Python 2.7. In `win_unicode_console` there is at least some support to get the Unicode argv and a possibility to have `sys.argv` automatically replaced, but it may easily break something depending on `sys.argv` being bytes.
